core/lb.go

```diff
@@ -0,0 +1,202 @@
+		lb.mu.Lock()
+		defer lb.mu.Unlock()
+		sess, exists := lb.sessions[job]
+		if !exists {
+			return
+		}
+		if sess.nonce != session.nonce {
+			// Session cached at `job` has since been recreated
+			return
+		}
+		delete(lb.sessions, job)
+		lb.load[transcoder] -= costEstimate
+		glog.V(common.DEBUG).Info("LB: Deleted transcode session for ", session.key)
+	}
+
+	go func() {
+		session.loop()
+		cleanupSession()
+	}()
+
+	glog.V(common.DEBUG).Info("LB: Created transcode session for ", session.key)
+	return session, nil
+}
+
+// Find the lowest loaded transcoder.
+// Expects the mutex `lb.mu` to be locked by the caller.
+func (lb *LoadBalancingTranscoder) leastLoaded() string {
+	min, idx := math.MaxInt64, 0
+	for i := 0; i < len(lb.transcoders); i++ {
+		k := (i + lb.idx) % len(lb.transcoders)
+		if lb.load[lb.transcoders[k]] < min {
+			min = lb.load[lb.transcoders[k]]
+			idx = k
+		}
+	}
+	return lb.transcoders[idx]
+}
+
+type transcoderParams struct {
+	job      string
+	fname    string
+	profiles []ffmpeg.VideoProfile
+	res      chan struct {
+		*TranscodeData
+		error
+	}
+}
+
+type transcoderSession struct {
+	transcoder LoadBalancedTranscoder
+	key        string
+	nonce      string
+
+	sender chan *transcoderParams
+}
+
+func (sess *transcoderSession) loop() {
+	defer func() {
+		sess.transcoder.Stop()
+		sess.sender = nil // unblocks any sender + select{ ... }
+	}()
+	// Run everything on a single loop to mitigate threading issues,
+	//   especially around transcoder cleanup
+	for {
+		ctx, cancel := transcodeLoopContext()
+		select {
+		case <-ctx.Done():
+			// Terminate the session after a period of inactivity
+			glog.V(common.DEBUG).Info("LB: Transcode loop timed out for ", sess.key)
+			return
+		case params := <-sess.sender:
+			cancel()
+			res, err :=
+				sess.transcoder.Transcode(params.job, params.fname, params.profiles)
+			params.res <- struct {
+				*TranscodeData
+				error
+			}{res, err}
+			if err != nil {
+				glog.V(common.DEBUG).Info("LB: Stopping transcoder due to error for ", sess.key)
+				return
+			}
+		}
+	}
+}
+
+func (sess *transcoderSession) Transcode(job string, fname string, profiles []ffmpeg.VideoProfile) (*TranscodeData, error) {
+	params := &transcoderParams{job: job, fname: fname, profiles: profiles,
+		res: make(chan struct {
+			*TranscodeData
+			error
+		})}
+	// XXX check transcoding on a looped out channel . May be threading issues...
+	select {
+	case sess.sender <- params:
+		glog.V(common.DEBUG).Info("LB: Transcode submitted for ", sess.key)
+	default:
+		glog.V(common.DEBUG).Info("LB: Transcoder was busy; exiting ", sess.key)
+		return nil, ErrTranscoderBusy
+	}
+	res := <-params.res
+	return res.TranscodeData, res.error
+}
+
+func calculateCost(profiles []ffmpeg.VideoProfile) int {
+	cost := 0
+	for _, v := range profiles {
+		w, h, err := ffmpeg.VideoProfileResolution(v)
+		if err != nil {
+			continue
+		}
+		cost += w * h * int(v.Framerate) // TODO incorporate duration
```

Comment:
Perhaps eventually - it would allow for more accurate load balancing - but that has the same issues mentioned here https://github.com/livepeer/go-livepeer/pull/1124#discussion_r344978928